src/osd/scrubber/scrub_backend.h

```diff
@@ -0,0 +1,419 @@
+  bool is_auth_available{false};  // managed to select an auth' source?
+  bool digest_match{true};        // do all (existing) digests match?
+};
+
+// note: some scrub tests are sensitive to the specific format of
+// auth_selection_t listing in the logs
+template <>
+struct fmt::formatter<auth_selection_t> {
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext& ctx)
+  {
+    return ctx.begin();
+  }
+
+  template <typename FormatContext>
+  auto format(auth_selection_t const& aus, FormatContext& ctx)
+  {
+    return fmt::format_to(ctx.out(),
+                          " {{AU-S: {}->{:x} OI({:x}:{}) {} dm:{}}} ",
+                          aus.auth->first,
+                          (uint64_t)(&aus.auth->second),
+                          (uint64_t)(&aus.auth_oi),
+                          aus.auth_oi,
+                          aus.shard_map.size(),
+                          aus.digest_match);
+  }
+};
+
+/**
+ * the back-end data that is per-chunk
+ *
+ * Created by the Scrubber after all replicas' maps have arrived.
+ */
+struct ScrubBeChunk {
+
+  explicit ScrubBeChunk(pg_shard_t i_am)
+  {
+    received_maps[i_am] = ScrubMap{};
+  }
+
+  /// the working set of scrub maps: the received maps, plus
+  /// Primary's own map.
+  std::map<pg_shard_t, ScrubMap> received_maps;
+
+  /// a collection of all objs mentioned in the maps
+  std::set<hobject_t> master_set;
+
+  utime_t started{ceph_clock_now()};
+
+  std::map<hobject_t, data_omap_digests_t> missing_digest;
+
+  /// Map from object with errors to good peers
+  std::map<hobject_t, std::list<pg_shard_t>> authoritative;
+
+
+  // these must be reset for each element:
+
+  std::set<pg_shard_t> cur_missing;
+  std::set<pg_shard_t> cur_inconsistent;
+  bool fix_digest{false};
+};
+
+
+/**
+ * ScrubBackend wraps the data and operations required for the back-end part of
+ * the scrubbing (i.e. for comparing the maps and fixing objects).
+ *
+ * Created anew upon each initiation of a scrub session.
+ *
+ * Uses 'optional' entries for winking-out the auxiliary collections.
+ */
+class ScrubBackend : public ScrubBackendIF {
+ public:
+  // Primary constructor
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep,
+               const std::set<pg_shard_t>& acting);
+
+  // Replica constructor: no primary map
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep);
+
+  friend class PgScrubber;
+
+  /**
+   * reset the per-chunk data structure (ScrubberBeChunk).
+   * Create an empty scrub-map for this shard, and place it
+   * in the appropriate entry in 'received_maps'.
+   *
+   * @returns a pointer to the newly created ScrubMap.
+   */
+  ScrubMap* new_chunk() final;
```

Comment:
It seems like PGScrubber::build_scrub_map() should probably return a ScrubMap that then gets passed here to something like ScrubBackend::add_primary_map.

Reply:
Was this way. Changed when I noticed that by having the primary map as just one more entry in the collected set of maps, I can avoid creating a "set of pointers to maps, which are all but one in the same vector" and then having to access the maps via this indirection (there is a comment somewhere to that effect).

Reply:
Seems like add_primary_map could still add to this set?

Reply:
Will create an extra copy. Prefer not to do that if no reason.

Reply:
Good point... will revise

Reply:
@athanatos - done